### PR TITLE
DataFixers, third times a charm with an incremental approach

### PR DESCRIPTION
**Pull request by i509VCB**
My last 2 DataFixer PRs had a major flaw: They were far too big. So this time I've decided to split up the process into more digestible bits. These pieces of the puzzle are the simplest to understand, so we start here.

This PR only includes:
A way for mods to register a datafixer for.
3 Simple fixes: Biome, Block and Item renames.

More advanced features will come in future PRs.

This supersedes #371 and #440

### CODE REVIEW

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/DataFixerHelper.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	/**
+	 * Gets the instance of the {@link DataFixerHelper}.
+	 */
+	DataFixerHelper INSTANCE = FabricDataFixerImpl.INSTANCE;
+
+	/**
+	 * Registers a DataFixer.
+	 *
+	 * @param modid The modid of the mod registering this DataFixer
+	 * @param runtimeDataVersion the current dataversion of the mod being ran.
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/DataFixerHelper.java]: how about `currentVersion`

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/DataFixerHelper.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ *
+ *     <p>Next your DataFixers would be registered onto the Schemas. It is advised to use {@link SimpleFixes} as that will cover most cases. Though you can create your own fixes which can be a complex process. How to use the fixes is shown in documentation of {@link SimpleFixes}
+ *
+ *     <p>Finally you need to build and register your DataFixer for use.
+ *         First build the DataFixer using {@link DataFixerBuilder#build(Executor)}. The executor should be {@link Util#getServerWorkerExecutor()}.
+ *         Finally register the DataFixer using {@link DataFixerHelper#registerFixer(String, int, DataFixer)} so that the DataFixer is invoked when the game normally fixes any data.
+ *
+ * <p>If you need more help, see the wiki article here: https://fabricmc.net/wiki/tutorial:datafixer or visit the Fabric Discord server.
+ */
+public interface DataFixerHelper {
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/DataFixerHelper.java]: like `FabricDataFixes`? don't think this name is super clear

**Review by liach — COMMENTED** [review]
Looks fairly good on first sight

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricSchemas.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * This allows quick access to common Schema types used in the game.
+ */
+public interface FabricSchemas {
+	/**
+	 * Fabric Schema Type. This is required for all custom DataFixers or fixing any data registered under {@link net.minecraft.datafixer.TypeReferences} will fail. This should be registered under version 0 within your DataFixer.
+	 */
+	BiFunction<Integer, Schema, Schema> FABRIC_SCHEMA = (version, parent) -> FabricDataFixesImpl.INSTANCE.fabricSchema;
+
+	/**
+	 * Normalize Identifier Schema.
```
Comment (Juuxel) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricSchemas.java:37]: This comment seems a bit pointless.

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricDataFixes.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	DataFixer registerFixer(String modid, int currentVersion, DataFixer dataFixer);
+
+	/**
+	 * Gets the DataFixer registered under a mod.
+	 *
+	 * @param modid The mod which the DataFixer was registered under.
+	 * @return An optional, which may contain a DataFixer if a mod has registered a DataFixer.
+	 * @throws IllegalArgumentException if the mod with the following id does not have a DataFixer.
+	 */
+	DataFixer getDataFixer(String modid);
```
Comment (Juuxel) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricDataFixes.java]:
```suggestion
	DataFixer getDataFixer(String modId);
```

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricDataFixes.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+public interface FabricDataFixes {
+	/**
+	 * Gets the instance of the {@link FabricDataFixes}.
+	 */
+	FabricDataFixes INSTANCE = FabricDataFixesImpl.INSTANCE;
+
+	/**
+	 * Registers a DataFixer.
+	 *
+	 * @param modid The modid of the mod registering this DataFixer
```
Comment (Juuxel) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricDataFixes.java]:
```suggestion
	 * @param modId The ID of the mod registering this DataFixer
```

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricDataFixes.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	 * @param currentVersion the current data version of the mod being ran.
+	 * @param dataFixer The DataFixer to register
+	 * @return The inputted DataFixer
+	 */
+	DataFixer registerFixer(String modid, int currentVersion, DataFixer dataFixer);
```
Comment (Juuxel) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricDataFixes.java]:
```suggestion
	DataFixer registerFixer(String modId, int currentVersion, DataFixer dataFixer);
```

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricDataFixes.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+	/**
+	 * Gets the DataFixer registered under a mod.
+	 *
+	 * @param modid The mod which the DataFixer was registered under.
```
Comment (Juuxel) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricDataFixes.java]:
```suggestion
	 * @param modId The mod which the DataFixer was registered under.
```

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricDataFixes.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	 * @return An optional, which may contain a DataFixer if a mod has registered a DataFixer.
+	 * @throws IllegalArgumentException if the mod with the following id does not have a DataFixer.
+	 */
+	DataFixer getDataFixer(String modid);
+
+	/**
+	 * Retrieves the DataVersion registered in a CompoundTag by a mod.
+	 *
+	 * @param compoundTag The CompoundTag to check
+	 * @param modid The id of the mod to check.
```
Comment (Juuxel) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricDataFixes.java]:
```suggestion
	 * @param modId The id of the mod to check.
```

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricDataFixes.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	 * @return The DataVersion stored for the mod or 0 if no DataVersion or mod is present.
+	 */
+	int getModDataVersion(CompoundTag compoundTag, String modid);
```
Comment (Juuxel) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/FabricDataFixes.java]:
```suggestion
	int getModDataVersion(CompoundTag compoundTag, String modId);
```

**Review by Juuxel — CHANGES_REQUESTED** [review]
Some fixes for the cursed capitalization of "mod ID".

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixesImpl.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+	public static final Logger LOGGER = LogManager.getLogger(FabricDataFixes.class);
+	public static FabricDataFixesImpl INSTANCE = new FabricDataFixesImpl();
+
+	public final Schema fabricSchema;
+	private final Map<String, DataFixerEntry> modDataFixers = new IdentityHashMap<>();
+	private boolean locked;
+
+	@Override
+	public DataFixer registerFixer(String modid, int currentVersion, DataFixer datafixer) {
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixesImpl.java:52]: Still need to replace `modid` with `modId` in this file

**Review by Juuxel — APPROVED** [review]

**Review by Ueaj-Kerman — APPROVED** [review]

**Review by liach — APPROVED** [review]
Remember to update to `fabric-datafixer-api-v1`

**Comment by i509VCB** [comment]
I need to find time to work on this soon lol

**Comment by i509VCB** [comment]
Going to close this PR and reopen in the future. There are more important pull requests that need to be attended to

**Pull request closed without merging**